### DISCUSSION TITLE
Chevrotain vs ANTLR4

### DISCUSSION THREAD

**Discussion by LNow** [eclipse-langium/langium#164] [General]
I'm quite new to creating DSL and I'm wondering what was the key features (or lack of feature) that convinced you to use Chevrotain over ANTLR4?

**Comment by msujew** [comment]
Hi @LNow, thank you for your question.

I think there were 3 main arguments in favor of Chevrotain compared to ANTLR4:

1. Chevrotain is way faster than ANTLR. You can run the benchmark for yourself over at [the chevrotain website](https://chevrotain.io/performance/). Be aware that it's optimized for the v8 engine (Chrome/Node.js).
2. Chevrotain does not need a second parser generator step. Currently, we generate a chevrotain parser from our grammar definition and that's it. Compiling that code to JavaScript is usually implicitely done by executing `tsc -w` on the whole project. ANTLR would need an additional process that then generates JavaScript code from our generated ANTLR grammar. This would also negatively impact our current `langium-cli` performance which comfortably sits at ~100ms (depending on the grammar complexity).
3. Ties in with the second argument: Keeping everything in the Node.js domain. ANTLR requires you to setup a JVM environment. It worked for Xtext, since it was JVM based. However, for Langium we try to keep external dependencies - especially runtimes - to a minimum.

Feature-wise, both ANTLR4 and Chevrotain offer us everything that we need, so there's no argument to be made.

There was also an internal argument generally against ANTLR4, since some colleagues argued that its error recovery performance degraded heavily compared to ANTLR3 (which is what Xtext currently uses). I didn't use ANTLR4 before, so I can only rely on their statements.
